File: 11_动态规划/子数组/152. 乘积最大子数组.ts
/**
 * @param {number[]} nums
 * @return {number}
 * @description 有点像 53. 最大子序和
 * 我们只要记录前i的最小值, 和最大值
 */
const maxProduct = function (nums: number[]): number {
  let res = nums[0]
  let preMax = nums[0]
  let preMin = nums[0]

  for (let i = 1; i < nums.length; i++) {
    const cur = nums[i]
    const curMax = Math.max(preMax * cur, preMin * cur, cur)
    const curMin = Math.min(preMax * cur, preMin * cur, cur)
    res = Math.max(res, curMax)
    preMax = curMax
    preMin = curMin
  }

  return res
}

console.log(maxProduct([2, 3, -2, 4]))

File: 11_动态规划/子数组/152. 乘积最大子数组.py
# 152. 乘积最大子数组
# https://leetcode.cn/problems/maximum-product-subarray/


from typing import List


def maxProduct(nums: List[int]) -> int:
    """
    乘积最大子数组.
    https://leetcode.cn/problems/maximum-product-subarray/
    """
    res, dpMin, dpMax = nums[0], nums[0], nums[0]
    for v in nums[1:]:
        ndpMin = min(dpMin * v, dpMax * v, v)
        ndpMax = max(dpMin * v, dpMax * v, v)
        dpMin, dpMax = ndpMin, ndpMax
        res = max(res, dpMax)
    return res


def maxStrength(nums: List[int]) -> int:
    """
    乘积最大子序列.
    https://leetcode.cn/problems/maximum-strength-of-a-group/
    """
    res, dpMin, dpMax = nums[0], nums[0], nums[0]
    for v in nums[1:]:
        ndpMin = min(dpMin * v, dpMax * v, v, dpMin)
        ndpMax = max(dpMin * v, dpMax * v, v, dpMax)
        dpMin, dpMax = ndpMin, ndpMax
        res = max(res, dpMax)
    return res


assert maxProduct([2, 3, -2, 4]) == 6
